### DISCUSSION TITLE
Issues interpolating current function and refining parameters

### DISCUSSION THREAD

**Discussion by Tyssel** [pybamm-team/PyBaMM#2344] [Q&A] [answered]
# Question 1
## Description
I want to use real-life data from a battery storage to model its batteries for synthetic data generation.
The approach for interpolation was taken from the drive_cycles.ipynb and looks like this in my case:

```python
timescale = param.evaluate(model.timescale)
current_function = pybamm.Interpolant(x=data.index.total_seconds().values,y=-data.current.values,children=timescale * pybamm.t, interpolator='cubic spline')
```
Where time is in 1-second steps, and current is given in Ampere. The latter has to be inverted, to match convention.

The current curve is shown below:

![image](https://user-images.githubusercontent.com/13622680/194014764-fa6ef5af-d32c-4a51-a1ab-1a76be92e2d4.png)

## Problem
The resulting interpolation just slightly resembles the given data, as seen below:

![image](https://user-images.githubusercontent.com/13622680/194016289-11e176b2-fd9c-495d-9fab-184115591246.png)

## current workaround
I have found, that multiplying the timescale term by some number >> 1 helps to match data much better, but I doubt that this is the best solution.

```python
current_function = pybamm.Interpolant(...,children=MULTIPLYER*timescale * pybamm.t, ...)
```
Which yields for MULTIPLYER = 10:

![image](https://user-images.githubusercontent.com/13622680/194018241-9f9a1a8f-49b2-4e56-8ca1-15bd4d53c5f5.png)

## Request

Is there any less improvised way to make the current interpolation better, or do I make a substantial mistake beforehand?

# Question 2
## Description
I know which cells are used in the battery storage, but the data sheet is very brief (I know it's a NMC(333)/C pouch cell, and its size), so I have to rely on PyBamm's parameter sets for most of the time.  The default solution is an OK approximation, but I have a pretty big dataset, so I thought I could try refining given parameter sets to my system. 

This is the result, for adapted electrode sizes:
![image](https://user-images.githubusercontent.com/13622680/194031675-a2d7fcc4-7162-4b3c-8569-36856767e7f8.png)

But it used a bad current interpolant:
![image](https://user-images.githubusercontent.com/13622680/194031881-ef09610f-980f-4fbf-ae26-a229adf3ec02.png)

My approach is the following:

- create load data
- create parameter grid (permutation of parameter ranges)
- create SPMe model
- load Mohtat2020 parameter set
- make discharge CCCV experiment until It reaches data[0] voltage and current reaches C/50 (as in #2263)
  - initialize model with conditions from previous solution
  - update parameters (one set from parameter grid)
  - solve
  - calculate MSE (V(data)-V(sim))
  - append params and MSE to list
- optimize parameter search with results list
- repeat

With this approach, I get parameter maps like this:
 
![image](https://user-images.githubusercontent.com/13622680/194031237-37f678f5-868b-43f3-ac7b-2c74e1981737.png)

## Problem
It is still not fully clear to me, which parameters are most relevant for the optimization. I cannot verify if my refinement uses valid models/current interpolant as I have to generate them on the fly. Current "best" parameter set makes it look worse:

```python
cathode_thickness_scale = 1.9
separator_thickness_scale = 2.1
windings = 40
temperature = 290
porosity = 0.5
```
![image](https://user-images.githubusercontent.com/13622680/194037446-b190d1c9-741c-4878-ba42-ca3a9586a118.png)

## currently changed/refined parameters
```python
            'Nominal cell capacity [A.h]' : 37,
            'Electrode height [m]': 0.188,
            'Electrode width [m]': 0.171*windings,
            'Negative electrode thickness [m]': anode_thickness_scale*param['Negative electrode thickness [m]'],
            'Positive electrode thickness [m]': cathode_thickness_scale*param['Positive electrode thickness [m]'],
            'Cell volume [m3]': 3.5e-4,
            'Upper voltage cut-off [V]': 4.2,
            'Lower voltage cut-off [V]': 2.5,
            'Number of cells connected in series to make a battery': 100,
            'Separator thickness [m]': separator_thickness_scale*param['Separator thickness [m]'],
            'Negative particle radius [m]': negative_particle_scaler * param['Negative particle radius [m]'],
            'Positive particle radius [m]': positive_particle_scaler * param['Positive particle radius [m]'],
            'Ambient temperature [K]': temperature,
            'Negative electrode active material volume fraction': anode_AM_vol,
            'Positive electrode active material volume fraction': cathode_AM_vol,
            'Positive electrode porosity': porosity,
            'Negative electrode porosity': porosity   
```
## request

- Is there a better way to optimize parameters by fitting?
- Are the parameters I've chosen the right ones?
- (What am I doing wrong?)

**Comment by YannickNoelStephanKuhn** [comment, marked as answer]
Fitting electrochemical models to data is a task big enough that there are plans to make a Python library for that. And even then, that can not work with a battery that may be arbitrarily different from the Mohtat2020 dataset.

You should work from what you know about your specific battery, without any of PyBaMM's standard parameter sets. You may use [https://liiondb.com/](https://liiondb.com/) for that. That will leave large gaps in your model parameter set, of course. Those you can systematically fill with the help of this paper:

[An asymptotic derivation of a single particle model with electrolyte](https://arxiv.org/abs/1905.12553)

You will see that the model parameters collapse into fewer non-dimensionalized ones. Any combination of parameters that generates the same set of non-dimensionalized parameters will look the same in the simulation.

**Reply by Tyssel** [reply]
Thank you for the links! That will give me a much better starting point.

Do you have any comments to the Current function interpolation? It seems to be also a rather arbitrary (and irreproducible) behaviour, which multiplier gives a good result.

**Reply by YannickNoelStephanKuhn** [reply]
You would need to label your axes: what is "Simulation"? Voltage? The current does not get simulated, after all. Maybe create a separate discussion for that, with a Minimum Working Example of the interpolation.

**Reply by Tyssel** [reply]
For completeness: It is current [A] vs. time [s].
I will open a new discussion for this.

**Reply by Tyssel** [reply]
@YannickNoelStephanKuhn It seems, that changing electrode sizes in the discharge_model, rather than the charge_model (which i was currently refining on) keeps things aligned.
